src/map/utils/battleutils.cpp

```diff
@@ -2229,14 +2229,84 @@ namespace battleutils
+            
+            //All zones from Adoulin onward have an id of 256+
+            //This includes Escha/Reisenjima and the new Dynamis zones
+            //(Not a post Adoulin Zone) && (Not Legion_A)
+            bool shouldApplyLevelCorrection = (zoneId < 256) && (zoneId != 183);
+
+            if(shouldApplyLevelCorrection) {
+                uint8 dLvl = PAttacker->GetMLevel() - PDefender->GetMLevel();
+                //Skip penalties for avatars, this should likely be all pets and mobs but I have no proof
+                //of this for ACC, ATT level correction for Pets/Avatars is the same as mobs though.
+                bool isPet = PAttacker->objtype == TYPE_PET;
+                bool isAvatar = false;
+                
+                if(isPet) {
+                    CPetEntity* petEntity = dynamic_cast<CPetEntity*>(PAttacker);
+                    isAvatar = petEntity->getPetType() == PETTYPE_AVATAR;
+                }
+
+                if(isAvatar)
+                {
+                    if(dLvl > 0)
+                    {
+                        //Avatars have a known level difference cap of 38
+                        hitrate += (int16)(std::min(dLvl, (uint8)38) * 2);
+                    }
+                }
+                else
+                {
+                    //Everything else has no known caps, though it's likely 38 like avatars
+                    hitrate += (int16)(dLvl * 2);
+                }
+            }
+
+            //https://www.bg-wiki.com/bg/Hit_Rate
+            //Update Notes: https://forum.square-enix.com/ffxi/threads/45365?p=534537#post534537
+            //The maximum accuracy of one-handed weapons equipped as the main weapon has been increased from 95% to 99%.
+            //* Owing to this change, the maximum accuracy of abilities that rely on main weapon accuracy has also been raised from 95% to 99%.
+            //Further, some monster damage types have been changed from hand-to-hand to blunt.* Fellows and alter egos enjoy this benefit as well.
+            //The maximum accuracy of beastmaster familiars, wyverns, avatars, and automatons has been increased from 95% to 99%.
+            //* In line with this change, familiars summoned using the following items have had their damage types changed from hand-to-hand to blunt.
+            //Carrot Broth / Famous Carrot Broth / Bug Broth / Quadav Bug Broth / Berbal Broth / Singing Herbal Broth / Carrion Broth / 
+            //Cold Carrion Broth / Meat Broth / Warm Meat Broth / Tree Sap / Scarlet Sap / Fish Broth / Fish Oil Broth / Seedbed Soil / Sun Water / 
+            //Grasshopper Broth / Noisy Grasshopper Broth / Mole Broth / Lively Mole Broth / Blood Broth / Clear Blood Broth / Antica Broth / Fragrant Antica Broth
+
+            int32 maxHitRate = 99;
+            auto targ_weapon = dynamic_cast<CItemWeapon*>(PAttacker->m_Weapons[SLOT_MAIN]);
```

Comment:
This cast can still produce `nullptr`, which will get used by `targ_weapon->isTwoHanded();` later, you can change that to:
`bool isTwoHanded = targ_weapon && targ_weapon->isTwoHanded();` to make it safe